File: CanliYayinJetpack/app/src/main/java/com/example/canliyayinjetpack/ViewerScreen.kt
package com.example.canliyayinjetpack

import android.graphics.BitmapFactory
import android.util.Base64
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.dp
import com.example.canliyayinjetpack.model.ChatMessageModel
import com.example.canliyayinjetpack.socket.LiveSocketListener
import com.example.canliyayinjetpack.socket.LiveSocketManager
import org.json.JSONObject

@Composable
fun ViewerScreen(
    roomId: String,
    roomTitle: String,
    onBackClick: () -> Unit
) {
    var statusText by remember { mutableStateOf("Bağlanıyor...") }
    var viewerCount by remember { mutableStateOf(0) }
    var messageText by remember { mutableStateOf("") }
    var chatMessages by remember { mutableStateOf<List<ChatMessageModel>>(emptyList()) }
    var liveBitmap by remember { mutableStateOf<android.graphics.Bitmap?>(null) }

    var socketManager by remember {
        mutableStateOf<LiveSocketManager?>(null)
    }

    DisposableEffect(roomId) {
        val manager = LiveSocketManager(
            serverUrl = AppConfig.SERVER_URL,
            listener = object : LiveSocketListener {

                override fun onConnected() {
                    statusText = "Sunucuya bağlandı, odaya giriliyor..."

                    val json = JSONObject()
                    json.put("type", "join_room")
                    json.put("room_id", roomId)
                    json.put("username", "Compose İzleyici")

                    socketManager?.sendJson(json)
                }

                override fun onMessage(message: String) {
                    val json = JSONObject(message)
                    val type = json.getString("type")

                    when (type) {
                        "joined_room" -> {
                            statusText = "Yayına bağlandı"
                        }

                        "viewer_count" -> {
                            viewerCount = json.getInt("viewer_count")
                        }

                        "chat_message" -> {
                            val chat = ChatMessageModel(
                                roomId = json.getString("room_id"),
                                username = json.getString("username"),
                                message = json.getString("message"),
                                createdAt = json.getString("created_at")
                            )

                            chatMessages = chatMessages + chat
                        }

                        "video_frame" -> {
                            val base64Frame = json.getString("frame")
                            liveBitmap = base64ToBitmap(base64Frame)
                        }

                        "stream_ended" -> {
                            statusText = "Yayın sona erdi"
                        }

                        "error" -> {
                            statusText = json.getString("message")
                        }
                    }
                }

                override fun onError(error: String) {
                    statusText = "Hata: $error"
                }

                override fun onDisconnected() {
                    statusText = "Bağlantı kapandı"
                }
            }
        )

        socketManager = manager
        manager.connect()

        onDispose {
            manager.disconnect()
        }
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Color.Black)
            .padding(12.dp)
    ) {

        Button(onClick = onBackClick) {
            Text("Geri")
        }

        Spacer(modifier = Modifier.height(8.dp))

        Text(
            text = roomTitle,
            color = Color.White,
            style = MaterialTheme.typography.titleLarge
        )

        Text(
            text = "İzleyici: $viewerCount",
            color = Color.White
        )

        Text(
            text = statusText,
            color = Color.White
        )

        Spacer(modifier = Modifier.height(8.dp))

        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(260.dp)
                .background(Color(0xFF111111))
        ) {
            liveBitmap?.let { bitmap ->
                Image(
                    bitmap = bitmap.asImageBitmap(),
                    contentDescription = "Canlı yayın görüntüsü",
                    modifier = Modifier.fillMaxSize(),
                    contentScale = ContentScale.Fit
                )
            }
        }

        Spacer(modifier = Modifier.height(8.dp))

        Text(
            text = "Canlı Sohbet",
            color = Color.White,
            style = MaterialTheme.typography.titleMedium
        )

        LazyColumn(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f)
        ) {
            items(chatMessages) { chat ->
                ChatMessageItem(chat)
            }
        }

        Row(
            modifier = Modifier.fillMaxWidth()
        ) {
            TextField(
                value = messageText,
                onValueChange = { messageText = it },
                modifier = Modifier.weight(1f),
                placeholder = {
                    Text("Mesaj yaz...")
                }
            )

            Spacer(modifier = Modifier.width(8.dp))

            Button(
                onClick = {
                    val msg = messageText.trim()

                    if (msg.isNotEmpty()) {
                        val json = JSONObject()
                        json.put("type", "chat_message")
                        json.put("message", msg)

                        socketManager?.sendJson(json)
                        messageText = ""
                    }
                }
            ) {
                Text("Gönder")
            }
        }
    }
}

@Composable
fun ChatMessageItem(chat: ChatMessageModel) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 4.dp)
    ) {
        Text(
            text = chat.username,
            color = Color(0xFF93C5FD),
            style = MaterialTheme.typography.labelMedium
        )

        Text(
            text = chat.message,
            color = Color.White
        )
    }
}

fun base64ToBitmap(base64Frame: String): android.graphics.Bitmap? {
    return try {
        val imageBytes = Base64.decode(base64Frame, Base64.DEFAULT)

        BitmapFactory.decodeByteArray(
            imageBytes,
            0,
            imageBytes.size
        )
    } catch (e: Exception) {
        null
    }
}

File: flutter_canli_yayin/lib/screens/viewer_screen.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:flutter/material.dart';

import '../config/app_config.dart';
import '../models/chat_message_model.dart';
import '../models/room_model.dart';
import '../services/live_socket_service.dart';

class ViewerScreen extends StatefulWidget {
  final RoomModel room;
  final VoidCallback onBack;

  const ViewerScreen({
    super.key,
    required this.room,
    required this.onBack,
  });

  @override
  State<ViewerScreen> createState() => _ViewerScreenState();
}

class _ViewerScreenState extends State<ViewerScreen> {
  LiveSocketService? socketService;

  String statusText = "Bağlanıyor...";
  int viewerCount = 0;
  Uint8List? currentFrame;

  final TextEditingController messageController = TextEditingController();
  final List<ChatMessageModel> chatMessages = [];

  @override
  void initState() {
    super.initState();
    viewerCount = widget.room.viewerCount;
    connectSocket();
  }

  void connectSocket() {
    socketService = LiveSocketService(
      serverUrl: AppConfig.serverUrl,

      onConnected: () {
        setState(() {
          statusText = "Sunucuya bağlandı, odaya giriliyor...";
        });

        socketService?.sendJson({
          "type": "join_room",
          "room_id": widget.room.roomId,
          "username": "Flutter İzleyici",
        });
      },

      onMessage: (message) {
        try {
          final data = jsonDecode(message);

          if (data["type"] == "joined_room") {
            setState(() {
              statusText = "Yayına bağlandı";
            });
          }

          if (data["type"] == "viewer_count") {
            setState(() {
              viewerCount = data["viewer_count"] ?? 0;
            });
          }

          if (data["type"] == "video_frame") {
            final base64Frame = data["frame"];

            setState(() {
              currentFrame = base64Decode(base64Frame);
            });
          }

          if (data["type"] == "chat_message") {
            final chat = ChatMessageModel.fromJson(data);

            setState(() {
              chatMessages.add(chat);
            });
          }

          if (data["type"] == "stream_ended") {
            setState(() {
              statusText = "Yayın sona erdi";
            });
          }

          if (data["type"] == "error") {
            setState(() {
              statusText = data["message"] ?? "Bilinmeyen hata";
            });
          }
        } catch (e) {
          setState(() {
            statusText = "Sunucudan gelen veri okunamadı";
          });
        }
      },

      onError: (error) {
        setState(() {
          statusText = "Hata: $error";
        });
      },

      onDisconnected: () {
        setState(() {
          statusText = "Bağlantı kapandı";
        });
      },
    );

    socketService?.connect();
  }

  void sendChatMessage() {
    final message = messageController.text.trim();

    if (message.isEmpty) return;

    socketService?.sendJson({
      "type": "chat_message",
      "message": message,
    });

    messageController.clear();
  }

  @override
  void dispose() {
    socketService?.disconnect();
    messageController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color(0xFF020617),
      appBar: AppBar(
        title: Text(widget.room.title),
        leading: IconButton(
          icon: const Icon(Icons.arrow_back),
          onPressed: widget.onBack,
        ),
      ),
      body: Column(
        children: [
          Padding(
            padding: const EdgeInsets.all(12),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  "İzleyici: $viewerCount",
                  style: const TextStyle(color: Colors.white),
                ),
                const SizedBox(height: 4),
                Text(
                  statusText,
                  style: const TextStyle(color: Colors.white70),
                ),
              ],
            ),
          ),

          Container(
            width: double.infinity,
            height: 260,
            margin: const EdgeInsets.symmetric(horizontal: 12),
            decoration: BoxDecoration(
              color: const Color(0xFF111827),
              borderRadius: BorderRadius.circular(12),
            ),
            clipBehavior: Clip.antiAlias,
            child: currentFrame == null
                ? const Center(
              child: Text(
                "Görüntü bekleniyor...",
                style: TextStyle(color: Colors.white54),
              ),
            )
                : Image.memory(
              currentFrame!,
              fit: BoxFit.contain,
              gaplessPlayback: true,
            ),
          ),

          const Padding(
            padding: EdgeInsets.all(12),
            child: Align(
              alignment: Alignment.centerLeft,
              child: Text(
                "Canlı Sohbet",
                style: TextStyle(
                  color: Colors.white,
                  fontWeight: FontWeight.bold,
                ),
              ),
            ),
          ),

          Expanded(
            child: ListView.builder(
              padding: const EdgeInsets.symmetric(horizontal: 12),
              itemCount: chatMessages.length,
              itemBuilder: (context, index) {
                final chat = chatMessages[index];

                return Padding(
                  padding: const EdgeInsets.only(bottom: 8),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        chat.username,
                        style: const TextStyle(
                          color: Color(0xFF93C5FD),
                          fontWeight: FontWeight.bold,
                          fontSize: 13,
                        ),
                      ),
                      Text(
                        chat.message,
                        style: const TextStyle(color: Colors.white),
                      ),
                    ],
                  ),
                );
              },
            ),
          ),

          Container(
            padding: const EdgeInsets.all(8),
            color: const Color(0xFF111827),
            child: Row(
              children: [
                Expanded(
                  child: TextField(
                    controller: messageController,
                    style: const TextStyle(color: Colors.white),
                    decoration: const InputDecoration(
                      hintText: "Mesaj yaz...",
                      hintStyle: TextStyle(color: Colors.white54),
                      filled: true,
                      fillColor: Color(0xFF1E293B),
                      border: OutlineInputBorder(),
                    ),
                  ),
                ),

                const SizedBox(width: 8),

                ElevatedButton(
                  onPressed: sendChatMessage,
                  child: const Text("Gönder"),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}